### PR TITLE
[NVIDIA GPU] Fix mem p2p init in collective permute thunk

### PR DESCRIPTION
**Pull request by Tixxx**
Move pointer initialization to the thunk init stage instead of runtime to get rid of the runtime blocking wait.
Add a device sync point using nccl allreduce before doing memcpy to make sure all gpus arrive at the same stage. Otherwise it's possible to have data corruptions when the receiving rank hasn't arrived at the memcpy.

### CODE REVIEW

**Review by frgossen — APPROVED** [review]
Thanks for the fix!

File: xla/service/gpu/runtime/nccl_collective_permute_thunk.cc [hunk 1]
```diff
@@ -161,13 +161,48 @@ absl::Status NcclCollectivePermuteStartThunk::Initialize(
...
+absl::Status NcclCollectivePermuteStartThunk::Cleanup(
+    const CleanupParams& params) {
+  TF_ASSIGN_OR_RETURN(const int64_t current_id,
+                      GetCurrentId(params.collective_params, config_));
+
+  absl::MutexLock lock(&barrier_mutex_);
+  if (!params.executor->HostMemoryUnregister(&barrier_flags_[current_id])) {
+    LOG(ERROR) << "Unregistering barrier flag failed.";
+  }
+}
```
Comment (thomasjoerg) [inline xla/service/gpu/runtime/nccl_collective_permute_thunk.cc:204]: Several TensorFlow tests are failing with:
`error: non-void function does not return a value in all control paths [-Werror,-Wreturn-type]`

**Review by thomasjoerg — CHANGES_REQUESTED** [review]

File: xla/service/gpu/runtime/nccl_collective_permute_thunk.cc (hunk 1 above)
Comment (Tixxx) [reply xla/service/gpu/runtime/nccl_collective_permute_thunk.cc:205]: added a returned status for the cleanup function

**Review by thomasjoerg — APPROVED** [review]

**Review by ddunl — APPROVED** [review]

**Comment by reedwm** [comment]
I'm seeing this regress the Maxtext LLama 7B model with 4-way FSDP and 2-way TP when collective matmul is enabled, using the script you gave me @Tixxx.  I see `Tokens/s/device: 4125.938` before this PR and `Tokens/s/device: 3827.136` after it.

**Pull request closed without merging**